File: android/app/src/main/java/com/sugarscratchie/smoke/data/SugarApi.kt
package com.sugarscratchie.smoke.data

import com.sugarscratchie.smoke.BuildConfig
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import kotlinx.serialization.Serializable
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.contentOrNull
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonPrimitive
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import java.net.URLEncoder
import java.util.concurrent.TimeUnit

class SugarApi(
    private val cookieJar: SessionCookieJar,
    private val apiBaseUrl: String = BuildConfig.API_BASE_URL.trimEnd('/'),
    private val mediaBaseUrl: String = BuildConfig.MEDIA_BASE_URL.trimEnd('/'),
) {
    private val json =
        Json {
            ignoreUnknownKeys = true
            isLenient = true
            encodeDefaults = true
        }

    private val client =
        OkHttpClient.Builder()
            .cookieJar(cookieJar)
            .connectTimeout(20, TimeUnit.SECONDS)
            .readTimeout(60, TimeUnit.SECONDS)
            .build()

    private val jsonMedia = "application/json; charset=utf-8".toMediaType()

    fun mediaUrl(path: String): String {
        if (path.startsWith("http://") || path.startsWith("https://")) return path
        var cleaned = path.trim()
        if (cleaned.startsWith("public/")) {
            cleaned = cleaned.removePrefix("public/")
        }
        if (!cleaned.startsWith("/")) {
            cleaned = "/$cleaned"
        }
        val encoded =
            cleaned.split("/").joinToString("/") { segment ->
                if (segment.isEmpty()) "" else URLEncoder.encode(segment, Charsets.UTF_8.name()).replace("+", "%20")
            }
        return "$mediaBaseUrl$encoded"
    }

    fun pickPlayableCard(cards: List<CardInfo>): CardInfo? =
        cards.firstOrNull {
            it.foreground.isNotBlank() && it.background.isNotBlank()
        } ?: cards.firstOrNull { it.foreground.isNotBlank() || it.background.isNotBlank() }

    fun backgroundUrl(card: CardInfo): String? {
        if (card.background.isBlank()) return null
        return mediaUrl(card.background)
    }

    fun foregroundUrl(card: CardInfo): String? {
        if (card.foreground.isBlank()) return null
        return mediaUrl(card.foreground)
    }

    suspend fun login(email: String, password: String): AuthResponse =
        postJson("/api/auth/login", CredentialsBody(email = email, password = password))

    suspend fun register(email: String, password: String): AuthResponse =
        postJson("/api/auth/register", CredentialsBody(email = email, password = password))

    suspend fun logout() {
        try {
            withContext(Dispatchers.IO) {
                val request =
                    Request.Builder()
                        .url("$apiBaseUrl/api/auth/logout")
                        .post("{}".toRequestBody(jsonMedia))
                        .header("Content-Type", "application/json")
                        .build()
                client.newCall(request).execute().use { /* ignore body */ }
            }
        } finally {
            cookieJar.clear()
        }
    }

    suspend fun session(): SessionResponse = get("/api/auth/session")

    suspend fun wallet(): WalletResponse = get("/api/me/wallet")

    suspend fun cards(): CardsResponse = get("/api/cards")

    suspend fun fetchBytes(url: String): ByteArray =
        withContext(Dispatchers.IO) {
            val request = Request.Builder().url(url).get().build()
            devMediaClient().newCall(request).execute().use { response ->
                if (!response.isSuccessful) {
                    throw ApiException(response.code, "HTTP ${response.code}")
                }
                response.body?.bytes() ?: ByteArray(0)
            }
        }

    suspend fun fetchText(url: String): String =
        withContext(Dispatchers.IO) {
            val request = Request.Builder().url(url).get().build()
            devMediaClient().newCall(request).execute().use { response ->
                val raw = response.body?.string().orEmpty()
                if (!response.isSuccessful) {
                    throw ApiException(response.code, "HTTP ${response.code}")
                }
                raw
            }
        }

    suspend fun startScratchHand(cardId: String): ScratchHandResponse =
        postJson("/api/rewards/scratch/hands", ScratchHandRequest(cardId = cardId))

    suspend fun claimScratchCoins(handId: String, milestone: Int, cardId: String): ScratchCoinsResponse =
        postJson(
            "/api/rewards/scratch/coins",
            ScratchCoinsRequest(handId = handId, milestone = milestone, cardId = cardId),
        )

    private suspend inline fun <reified T> get(path: String): T =
        withContext(Dispatchers.IO) {
            val request =
                Request.Builder()
                    .url("$apiBaseUrl$path")
                    .get()
                    .build()
            execute(request)
        }

    private suspend inline fun <reified Req, reified Res> postJson(path: String, body: Req): Res =
        withContext(Dispatchers.IO) {
            val encoded = json.encodeToString(body)
            val request =
                Request.Builder()
                    .url("$apiBaseUrl$path")
                    .post(encoded.toRequestBody(jsonMedia))
                    .header("Content-Type", "application/json")
                    .build()
            execute(request)
        }

    private inline fun <reified T> execute(request: Request): T {
        client.newCall(request).execute().use { response ->
            val raw = response.body?.string().orEmpty()
            if (!response.isSuccessful) {
                throw ApiException(response.code, parseDetail(raw) ?: "HTTP ${response.code}")
            }
            return json.decodeFromString(raw)
        }
    }

    private fun parseDetail(raw: String): String? {
        if (raw.isBlank()) return null
        return try {
            val err = json.decodeFromString<ApiErrorBody>(raw)
            val detail = err.detail ?: return raw
            when (detail) {
                is JsonPrimitive -> detail.contentOrNull ?: detail.toString()
                else -> {
                    try {
                        detail.jsonArray.joinToString("; ") { el ->
                            el.jsonPrimitive.contentOrNull ?: el.toString()
                        }
                    } catch (_: Exception) {
                        detail.toString()
                    }
                }
            }
        } catch (_: Exception) {
            raw.take(300)
        }
    }
}

@Serializable
private data class CredentialsBody(
    val email: String,
    val password: String,
)

File: ios/SugarScratchie/Data/SugarAPI.swift
import Foundation

final class SugarAPI: @unchecked Sendable {
    private let cookieStore: SessionCookieStore
    private let decoder = JSONDecoder()
    private let encoder = JSONEncoder()
    private let apiSession: URLSession

    init(cookieStore: SessionCookieStore) {
        self.cookieStore = cookieStore
        let config = URLSessionConfiguration.default
        config.timeoutIntervalForRequest = 8
        config.timeoutIntervalForResource = 20
        config.waitsForConnectivity = false
        config.httpCookieAcceptPolicy = .never
        config.httpShouldSetCookies = false
        self.apiSession = URLSession(configuration: config)
    }

    private var apiBaseUrl: String { DevEndpoints.apiBaseUrl }
    private var mediaBaseUrl: String { DevEndpoints.mediaBaseUrl }

    func mediaUrl(_ path: String) -> String {
        if path.hasPrefix("http://") || path.hasPrefix("https://") { return path }
        var cleaned = path.trimmingCharacters(in: .whitespacesAndNewlines)
        if cleaned.hasPrefix("public/") {
            cleaned = String(cleaned.dropFirst("public/".count))
        }
        if !cleaned.hasPrefix("/") {
            cleaned = "/" + cleaned
        }
        let encoded = cleaned
            .split(separator: "/", omittingEmptySubsequences: false)
            .map { segment -> String in
                if segment.isEmpty { return "" }
                return segment.addingPercentEncoding(withAllowedCharacters: .urlPathAllowed) ?? String(segment)
            }
            .joined(separator: "/")
        return mediaBaseUrl + encoded
    }

    func backgroundUrl(_ card: CardInfo) -> String? {
        guard !card.background.isEmpty else { return nil }
        return mediaUrl(card.background)
    }

    func foregroundUrl(_ card: CardInfo) -> String? {
        guard !card.foreground.isEmpty else { return nil }
        return mediaUrl(card.foreground)
    }

    func login(email: String, password: String) async throws -> AuthResponse {
        try await postJSON("/api/auth/login", body: CredentialsBody(email: email, password: password))
    }

    func register(email: String, password: String) async throws -> AuthResponse {
        try await postJSON("/api/auth/register", body: CredentialsBody(email: email, password: password))
    }

    func logout() async {
        defer { cookieStore.clear() }
        do {
            var request = URLRequest(url: URL(string: "\(apiBaseUrl)/api/auth/logout")!)
            request.httpMethod = "POST"
            request.setValue("application/json", forHTTPHeaderField: "Content-Type")
            request.httpBody = Data("{}".utf8)
            cookieStore.apply(to: &request)
            let (_, response) = try await apiSession.data(for: request)
            cookieStore.capture(from: response)
        } catch {
            // ignore — cookie cleared in defer
        }
    }

    func session() async throws -> SessionResponse {
        try await get("/api/auth/session")
    }

    func wallet() async throws -> WalletResponse {
        try await get("/api/me/wallet")
    }

    func cards() async throws -> CardsResponse {
        try await get("/api/cards")
    }

    func startScratchHand(cardId: String) async throws -> ScratchHandResponse {
        try await postJSON("/api/rewards/scratch/hands", body: ScratchHandRequest(cardId: cardId))
    }

    func claimScratchCoins(handId: String, milestone: Int, cardId: String) async throws -> ScratchCoinsResponse {
        try await postJSON(
            "/api/rewards/scratch/coins",
            body: ScratchCoinsRequest(handId: handId, milestone: milestone, cardId: cardId)
        )
    }

    func fetchBytes(_ urlString: String) async throws -> Data {
        guard let url = URL(string: urlString) else {
            throw ApiException(statusCode: 400, message: "Bad URL")
        }
        let (data, response) = try await DevMediaClient.shared.data(from: url)
        guard let http = response as? HTTPURLResponse else {
            throw ApiException(statusCode: 0, message: "No HTTP response")
        }
        guard (200..<300).contains(http.statusCode) else {
            throw ApiException(statusCode: http.statusCode, message: "HTTP \(http.statusCode)")
        }
        return data
    }

    func fetchText(_ urlString: String) async throws -> String {
        let data = try await fetchBytes(urlString)
        return String(decoding: data, as: UTF8.self)
    }

    /// Download remote media to a cached file so AVPlayer can play Vite basic-ssl URLs.
    func cachedMediaFile(urlString: String) async throws -> URL {
        guard let remote = URL(string: urlString) else {
            throw ApiException(statusCode: 400, message: "Bad media URL")
        }
        let caches = FileManager.default.urls(for: .cachesDirectory, in: .userDomainMask).first!
        let folder = caches.appendingPathComponent("sugar_media", isDirectory: true)
        try FileManager.default.createDirectory(at: folder, withIntermediateDirectories: true)
        let name = remote.path
            .replacingOccurrences(of: "/", with: "_")
            .replacingOccurrences(of: " ", with: "_")
        let dest = folder.appendingPathComponent(name.isEmpty ? UUID().uuidString : name)
        if FileManager.default.fileExists(atPath: dest.path) {
            return dest
        }
        let data = try await fetchBytes(urlString)
        try data.write(to: dest, options: .atomic)
        return dest
    }

    private func get<T: Decodable>(_ path: String) async throws -> T {
        var request = URLRequest(url: URL(string: "\(apiBaseUrl)\(path)")!)
        request.httpMethod = "GET"
        cookieStore.apply(to: &request)
        return try await execute(request)
    }

    private func postJSON<Req: Encodable, Res: Decodable>(_ path: String, body: Req) async throws -> Res {
        var request = URLRequest(url: URL(string: "\(apiBaseUrl)\(path)")!)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.httpBody = try encoder.encode(body)
        cookieStore.apply(to: &request)
        return try await execute(request)
    }

    private func execute<T: Decodable>(_ request: URLRequest) async throws -> T {
        let (data, response) = try await apiSession.data(for: request)
        cookieStore.capture(from: response)
        guard let http = response as? HTTPURLResponse else {
            throw ApiException(statusCode: 0, message: "No HTTP response")
        }
        if !(200..<300).contains(http.statusCode) {
            throw ApiException(statusCode: http.statusCode, message: parseDetail(data) ?? "HTTP \(http.statusCode)")
        }
        return try decoder.decode(T.self, from: data)
    }

    private func parseDetail(_ data: Data) -> String? {
        if data.isEmpty { return nil }
        if let err = try? decoder.decode(ApiErrorBody.self, from: data),
           let message = err.detail?.message
        {
            return message
        }
        return String(decoding: data.prefix(300), as: UTF8.self)
    }
}

private struct CredentialsBody: Codable {
    var email: String
    var password: String
}
